File: sol/solution/0101-0200/0105/Solution.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    public TreeNode BuildTree(int[] preorder, int[] inorder) {        
        Dictionary<int, int> dict = new Dictionary<int, int>();
        for (int i = 0; i < inorder.Length; i++) dict[inorder[i]] = i;
        return BuildTreeNode(0, preorder.Length - 1, 0, inorder.Length - 1);
        TreeNode BuildTreeNode(int preStartIdx, int preEndIdx, int inStartIdx, int inEndIdx){
            if (preStartIdx > preEndIdx || inStartIdx > inEndIdx) return null;
            int root = preorder[preStartIdx];
            TreeNode node = new TreeNode(root);
            int InorderIdx = dict[root];
            int InorderLeftLen = InorderIdx - inStartIdx;
            node.left = BuildTreeNode(preStartIdx + 1, preStartIdx + InorderLeftLen, inStartIdx, InorderIdx - 1);
            node.right = BuildTreeNode(preStartIdx + InorderLeftLen + 1, preEndIdx, InorderIdx + 1, inEndIdx);
            return node;
        }
    }
}

File: sol/solution/0101-0200/0105/Solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* buildTreeHelper(vector<int>& preorder, int prestart, int preend, vector<int>& inorder, int instart, int inend, unordered_map<int, int>& mpp) {
        if (prestart > preend || instart > inend) return nullptr;
        TreeNode* root = new TreeNode(preorder[prestart]);
        int inRootIndex = mpp[root->val];
        int numsLeft = inRootIndex - instart;
        root->left = buildTreeHelper(preorder, prestart + 1, prestart + numsLeft, inorder, instart, inRootIndex - 1, mpp);
        root->right = buildTreeHelper(preorder, prestart + numsLeft + 1, preend, inorder, inRootIndex + 1, inend, mpp);
        return root;
    }
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        unordered_map<int, int> mpp;
        for (int i = 0; i < inorder.size(); i++) 
            mpp[inorder[i]] = i;  
        return buildTreeHelper(preorder, 0, preorder.size() - 1, inorder, 0, inorder.size() - 1, mpp);
    }
};